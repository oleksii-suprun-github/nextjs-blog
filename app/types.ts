export type SanityCategory = {
  title?: string;
  description?: string;
};

export type SanityPostMeta = {
  title: string;
};

type SanityPostBase = {
  title: string;
  category: {
    title: string;
    url: string;
  };
  slug: string;
  image: {
    url: string;
    alt?: string;
  };
  createdAt: string;
};

export type SanityPostPreview = SanityPostBase & {
  previewText: string;
};

export type SanityPost = SanityPostBase & {
  content: string;
};
