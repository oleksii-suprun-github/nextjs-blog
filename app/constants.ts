export const SITE_BRAND_TITLE_ENDING = '| NextJS Sample Blog';

export const NEXTJS_URL = '/publications/nextjs';

export const NODEJS_URL = '/publications/nodejs';

export const PYTHON_URL = '/publications/python';
