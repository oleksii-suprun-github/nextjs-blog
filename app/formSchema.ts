import { z } from 'zod';

export const formSchema = z.object({
  name: z
    .string()
    .min(2, { message: 'Name must be at least 2 characters long' })
    .max(50, { message: 'Name must be less than 50 characters' }),
  email: z.string().email({ message: 'Please enter a valid email address' }),
  message: z
    .string()
    .min(10, { message: 'Message must be at least 10 characters long' })
    .max(500, { message: 'Message must be less than 500 characters' }),
});
