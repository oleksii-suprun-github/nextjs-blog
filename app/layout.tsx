import { SpeedInsights } from '@vercel/speed-insights/next';
import './globals.css';

export default function RootLayout({ children }: { children: React.ReactNode }) {
  return (
    <html className="scroll-smooth" lang="en">
      <body className={'min-h-dvh text-stone-300'}>
        {children}
        <SpeedInsights />
      </body>
    </html>
  );
}
