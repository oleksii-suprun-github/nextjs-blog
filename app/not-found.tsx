import NotFound from './components/NotFound';

export const metadata = {
  title: 'Page Not Found | Agora Energiewende',
  description: 'Sorry, we can`t find that publication.',
};

const PageNotFound = () => <NotFound link="/publications" />;

export default PageNotFound;
