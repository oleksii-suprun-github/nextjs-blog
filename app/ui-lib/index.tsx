export { default as Loading } from './Loading';
export { default as HeadlineWithDescription } from './HeadlineWithDescription';
export { default as PublicationsSection } from './PublicationsSection';
export { default as PublicationCard } from './PublicationCard';
export { default as PostHero } from './PostHero';
export { default as Form } from './Form';
export { default as FormInput } from './FormInput';
export { default as Modal } from './Modal';
export { default as Drawer } from './Drawer';
