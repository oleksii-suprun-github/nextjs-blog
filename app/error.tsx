'use client';

import { useEffect } from 'react';
import ErrorMessage from './components/ErrorMessage';
import { SITE_BRAND_TITLE_ENDING } from './constants';

export const metadata = {
  title: `Something went wrong ${SITE_BRAND_TITLE_ENDING}`,
};

export default function GlobalError({
  error,
  reset,
}: {
  error: Error & { digest?: string };
  reset: () => void;
}) {
  useEffect(() => {
    console.error('An unexpected error occurred:', error);
  }, [error]);

  return (
    <html>
      <body>
        <ErrorMessage error={error} reset={reset} />
      </body>
    </html>
  );
}
