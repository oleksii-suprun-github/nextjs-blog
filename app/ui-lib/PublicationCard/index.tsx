import Image from 'next/image';
import Link from 'next/link';
import { SanityPostPreview } from '@/app/types';

function PublicationCard({
  title,
  category,
  previewText,
  slug,
  image,
  publishedAt,
}: SanityPostPreview) {
  const link = `/publications/${category.url}/${slug}`;
  const creationDate = new Intl.DateTimeFormat('en-GB', {
    dateStyle: 'long',
  }).format(new Date(publishedAt));

  return (
    <Link href={link}>
      <div className="card card-compact bg-base-100 shadow-xl">
        <figure className="lxs-sm:h-[200px] md:h-[250px]">
          <Image
            width={400}
            height={400}
            src={image.url}
            alt={image.alt || title}
            quality={60}
            placeholder="blur"
            blurDataURL="iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNkmHOiHgAD7QHlxT90/wAAAABJRU5ErkJggg=="
            priority
            className="h-full w-full object-cover"
          />
        </figure>
        <div className="card-body sm:h-[340px] md:h-[320px] lg:h-[265px]">
          <span className="text-md text-slate-300">{creationDate}</span>
          <h2 className="card-title mb-4 text-stone-200">{title}</h2>
          <p className="text-stone-200">{previewText}</p>
          <div className="card-actions justify-end">
            <button className="btn no-animation btn-sm bg-brand-dark-purple text-gray-300 hover:bg-brand-purple">
              Read More
            </button>
          </div>
        </div>
      </div>

      {/* <div className="card image-full shadow-xl">
        <figure className="sm:h-[470px] lxs-sm:h-[380px] md:h-[420px]">
          <Image
            width={400}
            height={400}
            src={image.url}
            alt={image.alt || title}
            quality={60}
            placeholder="blur"
            blurDataURL="iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAADUlEQVR42mNkmHOiHgAD7QHlxT90/wAAAABJRU5ErkJggg=="
            priority
            className="h-full w-full object-cover"
          />
        </figure>
        <div className="card-body">
          <span className="text-md text-slate-300">{creationDate}</span>
          <h2 className="card-title mb-4 text-stone-200">{title}</h2>
          <p className="text-stone-200">{previewText}</p>
          <div className="card-actions justify-end">
            <button className="btn no-animation btn-sm bg-brand-dark-purple text-gray-300 hover:bg-brand-purple">
              Read More
            </button>
          </div>
        </div>
      </div> */}
    </Link>
  );
}

export default PublicationCard;
