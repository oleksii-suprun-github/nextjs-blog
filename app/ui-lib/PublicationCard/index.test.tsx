import { render, screen } from '@testing-library/react';
import PublicationCard from '.';
import { describe, it, expect, vi } from 'vitest';
import { SanityPostPreview } from '@/app/types';

// Mock the urlFor function from Sanity
vi.mock('@/sanity/lib/image', () => ({
  urlFor: () => ({
    width: () => ({
      url: () => '/mock-image-url.jpg',
    }),
  }),
}));

const mockPost: SanityPostPreview = {
  title: 'Test Title',
  category: { title: 'Test Category', url: 'test-category' },
  previewText: 'This is a preview of the post.',
  slug: 'test-slug',
  image: {
    url: '/test-image.jpg',
    alt: 'Test Image',
  },
  publishedAt: '2024-10-12T12:00:00Z',
};

describe('PublicationCard Component', () => {
  it('should render correctly and match snapshot', () => {
    const { asFragment } = render(<PublicationCard {...mockPost} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render the title, preview text, and creation date', () => {
    render(<PublicationCard {...mockPost} />);

    const titleElement = screen.getByText('Test Title');
    const previewTextElement = screen.getByText('This is a preview of the post.');
    const creationDateElement = screen.getByText('12 October 2024');

    expect(titleElement).toBeInTheDocument();
    expect(previewTextElement).toBeInTheDocument();
    expect(creationDateElement).toBeInTheDocument();
  });

  it('should render the image with correct attributes', () => {
    render(<PublicationCard {...mockPost} />);

    const imageElement = screen.getByAltText('Test Image');
    expect(imageElement).toBeInTheDocument();
    expect(imageElement).toHaveAttribute('src', expect.stringContaining('mock-image-url.jpg'));
    expect(imageElement).toHaveAttribute('width', '400');
    expect(imageElement).toHaveAttribute('height', '400');
  });

  it('should render the link with the correct href', () => {
    render(<PublicationCard {...mockPost} />);

    const linkElement = screen.getByRole('link');
    expect(linkElement).toBeInTheDocument();
    expect(linkElement).toHaveAttribute('href', '/publications/test-category/test-slug');
  });

  it('should render the Read More button', () => {
    render(<PublicationCard {...mockPost} />);

    const buttonElement = screen.getByRole('button', { name: 'Read More' });
    expect(buttonElement).toBeInTheDocument();
  });
});
