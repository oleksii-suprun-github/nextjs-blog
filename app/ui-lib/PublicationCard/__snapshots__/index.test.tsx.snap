// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PublicationCard Component > should render correctly and match snapshot 1`] = `
<DocumentFragment>
  <a
    href="/publications/test-category/test-slug"
  >
    <div
      class="card card-compact bg-base-100 shadow-xl"
    >
      <figure
        class="lxs-sm:h-[200px] md:h-[250px]"
      >
        <img
          alt="Test Image"
          class="h-full w-full object-cover"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="400"
          src="/_next/image?url=%2Ftest-image.jpg&w=828&q=60"
          srcset="/_next/image?url=%2Ftest-image.jpg&w=640&q=60 1x, /_next/image?url=%2Ftest-image.jpg&w=828&q=60 2x"
          style="color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
          width="400"
        />
      </figure>
      <div
        class="card-body sm:h-[340px] md:h-[320px] lg:h-[265px]"
      >
        <span
          class="text-md text-slate-300"
        >
          12 October 2024
        </span>
        <h2
          class="card-title mb-4 text-stone-200"
        >
          Test Title
        </h2>
        <p
          class="text-stone-200"
        >
          This is a preview of the post.
        </p>
        <div
          class="card-actions justify-end"
        >
          <button
            class="btn no-animation btn-sm bg-brand-dark-purple text-gray-300 hover:bg-brand-purple"
          >
            Read More
          </button>
        </div>
      </div>
    </div>
  </a>
</DocumentFragment>
`;
